### PR TITLE
Add limit for active password resets

### LINKED ISSUE #155: SECURITY: Add limit on pending password reset emails
A malicious agent could easily visit the reset password page, fill in a email and press "send reset link". At least at a rate of 3600 messages/hour. The *forgot password* view should only approve of sending an email if the following conditions are met: **No more than 1 pending (active) password reset exists within the last 2 hours for the target username**

In effect, this means a really evil agent could spam a user at a rate of 24 messages per day; fairly manageable. But not like >86 k emails/day.

### PR DESCRIPTION
**Pull request by mxtthias**
Closes #155.

### CODE REVIEW

**Comment by mxtthias** [comment]
@mblomdahl Not sure if I've put the logic in the right places, but I figure it's a start.

File: tests/end2end/test_public_reset_password.py [hunk 1]
```diff
@@ -141,3 +141,42 @@ def test_sees_error_message_if_attempting_to_use_reset_code_twice(password_reset
     # Sees error.
     assert escape(_('Reset code "%(code)s" already used (%(isoformat)s)', code=password_reset.code,
                     isoformat=password_reset.modified_at.isoformat() + 'Z')) in res
+
+
+def test_sees_error_message_if_too_many_active_password_resets(db, testapp):
+    """Show error if too many active password resets exist."""
+    inactive_user = UserFactory(is_active=False)
```
Comment (mblomdahl) [inline tests/end2end/test_public_reset_password.py]: I think it's a bit confusing explicitly making the user inactive for this test case. Makes one wonder, why so?

Easier to understand the test case if it contains the bare minimum for satisfying a requirement; "error message instead of email if too many active (and also recent) password resets are pending".

(In the test case this was copy-pasted from, the purpose of it being inactive (initially) was to assert that after completing the request reset -> click link -> fill in new password workflow, the account would change state to active.)

Proposing to ...
1. add any user via the arg fixture, `..._active_password_resets(user, testapp)`
2. remove stuff related to user being active/inactive

File: tests/end2end/test_public_reset_password.py [hunk 2]
```diff
@@ -141,3 +141,42 @@ def test_sees_error_message_if_attempting_to_use_reset_code_twice(password_reset
...
+    assert inactive_user.is_active is False
+
+    # Create active password resets
+    for _i in range(0, PasswordReset.MAX_ALLOWED_ACTIVE_PASSWORD_RESETS):
+        # Goes to homepage.
+        res = testapp.get('/')
+        # Clicks on 'Forgot password'.
+        res = res.click(_('Forgot password?'))
+        # Fills out ForgotPasswordForm.
+        username_with_different_casing = inactive_user.email.upper()
```
Comment (mblomdahl) [inline tests/end2end/test_public_reset_password.py]: I think this "works irrespective of casing" logic has already been tested in other test case. It appears as clutter here, proposing to replace it with the simpler version used in many other tests:

```python
        # Fills out ForgotPasswordForm.
        form = res.forms['forgotPasswordForm']
        form['username'] = user.email
```

File: tests/end2end/test_public_reset_password.py [hunk 3]
```diff
@@ -141,3 +141,42 @@ def test_sees_error_message_if_attempting_to_use_reset_code_twice(password_reset
...
+    active_resets = PasswordReset.get_active_resets_for_email(inactive_user.email)
+    assert len(active_resets) == PasswordReset.MAX_ALLOWED_ACTIVE_PASSWORD_RESETS
+
+    # Try to create one additional password reset
+    # Goes to homepage.
+    res = testapp.get('/')
+    # Clicks on 'Forgot password'.
+    res = res.click(_('Forgot password?'))
+    # Fills out ForgotPasswordForm.
+    username_with_different_casing = inactive_user.email.upper()
```
Comment (mblomdahl) [inline tests/end2end/test_public_reset_password.py]: See previous comment!

File: xl_auth/public/forms.py [hunk 4]
```diff
@@ -61,7 +61,16 @@ def validate(self):
 
         user = User.get_by_email(self.username.data)
         if user:
-            return True
+            active_resets = PasswordReset.get_active_resets_for_email(user.email)
```
Comment (mblomdahl) [inline xl_auth/public/forms.py]: I think this is a step away from the previous approach where we have *users* and *collections* at the center. Each of them relates to a bunch of other stuffs, like permissions and clients and tokens and other users and roles and password resets. When not resolved directly from the relation, they have utility methods to filter among related entities. E.g.

```python
class User(UserMixin, SurrogatePK, Model):
    """A user of the app."""
    # ...

    permissions = relationship('Permission', back_populates='user',
                               foreign_keys='Permission.user_id', lazy='joined')

    # ...

    def has_any_permission_for(self, collection):
        """Check for any permission on a specific collection."""
        return any([permission for permission in self.permissions
                    if permission.collection == collection])

    # ...

    # ...

    def get_cataloging_admin_permissions(self):
        """Return all cataloging admin permissions for this user."""
        return [perm for perm in self.permissions if perm.cataloging_admin]
```

I think a simpler and better maintainable approach would be adding this logic alongside *get_subset_of_permission* and friends. E.g.

```python
class PasswordReset(SurrogatePK, Model):
    """Password reset token for a user."""

   # ...

    created_at = Column(db.DateTime, default=datetime.utcnow, nullable=False)

    # ...

    @hybrid_property
    def is_recent(self):
        """Check if reset was recently created."""
        return self.created_at > (datetime.utcnow() - timedelta(hours=2))

# ...

class User(UserMixin, SurrogatePK, Model):
    """A user of the app."""

    # ...

    password_resets = relationship('PasswordReset', back_populates='user')
    
    # ...

    def get_active_and_recent_password_resets(self):
        """Return active and recent password resets for the user."""
        return list(filter(lambda _: _.is_active and _.is_recent, self.password_resets))

```

This way, form logic could be more like

```python
        # ...
        if user:
            if len(user.get_active_and_recent_password_resets()) >= 2:
                self.username.errors.append(_('Chill down and wait')
                return False
            else:
                return True
```

Also means neater and better unittests for the models, methinks.

What do you think?

File: tests/end2end/test_public_reset_password.py (hunk 1 above)
Comment (mxtthias) [reply tests/end2end/test_public_reset_password.py]: Copy-paste from `test_can_complete_password_reset_flow(db, testapp)`. I'll clean it up!

File: tests/end2end/test_public_reset_password.py (hunk 2 above)
Comment (mxtthias) [reply tests/end2end/test_public_reset_password.py]: :+1:

File: tests/end2end/test_public_reset_password.py (hunk 3 above)
Comment (mxtthias) [reply tests/end2end/test_public_reset_password.py]: :+1:

**Comment by mblomdahl** [comment]
@mxtthias I was very slow with merging 0.7.4 release, when I did this morning it introduced a pretty massive merge conflict. Now resolved here with e3fff03. Feel free squashing that one away too. :)

File: xl_auth/user/models.py [hunk 5]
```diff
@@ -179,6 +184,8 @@ class User(UserMixin, SurrogatePK, Model):
     created_by_id = reference_col('users', nullable=False)
     created_by = relationship('User', remote_side=id, foreign_keys=created_by_id)
 
+    MAX_ALLOWED_ACTIVE_PASSWORD_RESETS = 2
```
Comment (mblomdahl) [inline xl_auth/user/models.py]: I think *User. MAX_ALLOWED_ACTIVE_PASSWORD_RESETS* is [a bit at odds with the rest of the code base](https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds), for a number of reasons.

First is we don't use constants anywhere else in the models; e.g. the number of seconds that a grant/access token is valid, what amount of time constitutes "recent", how long before a password reset expires, and so forth.

Second is that it (in my eyes) kinda gives the impression that it's something that could (should?) be changed at will. But in the test case layout is such that we in end2end test for *assert password_resets <= User.MAX_PASSWORD_...* and in model-unit we test for what is specified by #155 (exactly two).

Third thing noticed is that if this should be implemented as a "possibly maybe future configurable", then it would be more "at home" alongside *BCRYPT_LOG_ROUNDS*, *CACHE_TYPE* and friends in `xl_auth.settings.py`.

So, bottom line, proposing to ...
1. replace constant `User.MAX_ALLOWED_ACTIVE_PASSWORD_RESETS` with `2` in the 1 place that the Flask app accesses it + tag the squash-commit with `#155` so it's easy to see where it comes from on git-blame, OR
2. make `MAX_ALLOWED_ACTIVE_PASSWORD_RESETS` into an app config setting (not necessarily overridable at this point)

That's my two cents! But aesthetics apart; LGTM.
